diff --git a/include/libtorrent/aux_/torrent.hpp b/include/libtorrent/aux_/torrent.hpp
index fbbd81fcd028aa0a0feb0722da0427a87efc28ac..285fc5dcc33f82da17e68cec6927270fff9bd8a7 100644
--- a/include/libtorrent/aux_/torrent.hpp
+++ b/include/libtorrent/aux_/torrent.hpp
@@ -1657,11 +1657,13 @@ namespace libtorrent::aux {
 		// set_error()
 		file_index_t m_error_file;
 
+#ifndef TORRENT_DISABLE_STREAMING
 		// the average time it takes to download one time critical piece
 		std::int32_t m_average_piece_time = 0;
 
 		// the average piece download time deviation
 		std::int32_t m_piece_time_deviation = 0;
+#endif // TORRENT_DISABLE_STREAMING
 
 		// the number of bytes that has been
 		// downloaded that failed the hash-test
